fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,357 @@
+            state_tree.set_actor(i, state);
+        }
+        let root_cid = state_tree.flush().unwrap();
+        (root_cid, state_tree)
+    }
+
+    fn assert_tree2_contains_tree1<Store1: Blockstore, Store2: Blockstore>(
+        tree1: &StateTree<Store1>,
+        tree2: &StateTree<Store2>,
+    ) {
+        tree1
+            .for_each(|addr, state| {
+                let r = tree2.get_actor_by_address(&addr);
+                if r.is_err() {
+                    panic!("addr: {addr:?} does not exists in tree 2");
+                }
+
+                if let Some(target_state) = r.unwrap() {
+                    assert_eq!(target_state, *state);
+                } else {
+                    panic!("missing address: {addr:?}");
+                }
+                Ok(())
+            })
+            .unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_streamer() {
+        let (root_cid, state_tree) = prepare_state_tree(100);
+        let bs = state_tree.into_store();
+        let mut stream = StateTreeStreamer {
+            dfs: VecDeque::from(vec![root_cid.clone()]),
+            bs: bs.clone(),
+        };
+
+        let new_bs = MemoryBlockstore::new();
+        while let Some((cid, bytes)) = stream.next().await {
+            new_bs.put_keyed(&cid, &bytes).unwrap();
+        }
+
+        let new_state_tree = StateTree::new_from_root(new_bs, &root_cid).unwrap();
+        let old_state_tree = StateTree::new_from_root(bs, &root_cid).unwrap();
+
+        assert_tree2_contains_tree1(&old_state_tree, &new_state_tree);
+        assert_tree2_contains_tree1(&new_state_tree, &old_state_tree);
+    }
+
+    #[tokio::test]
+    async fn test_car() {
+        let (state_root, state_tree) = prepare_state_tree(100);
+        let state_params = FvmStateParams {
+            state_root,
+            timestamp: Timestamp(100),
+            network_version: NetworkVersion::V1,
+            base_fee: Default::default(),
+            circ_supply: Default::default(),
+            chain_id: 1024,
+        };
+        let block_height = 2048;
+
+        let bs = state_tree.into_store();
+        let db = ReadOnlyBlockstore::new(bs.clone());
+        let snapshot = Snapshot::new(db, state_params.clone(), block_height).unwrap();
+
+        let tmp_file = tempfile::NamedTempFile::new().unwrap();
+        let r = snapshot.write_car(tmp_file.path()).await;
+        assert!(r.is_ok());
+
+        let new_store = MemoryBlockstore::new();
+        let loaded_snapshot = match Snapshot::read_car(tmp_file.path(), new_store)
+            .await
+            .unwrap()
+        {
+            Snapshot::V1(s) => s,
+        };
+
+        assert_eq!(state_root, loaded_snapshot.state_tree_root);
+        assert_tree2_contains_tree1(
+            &StateTree::new_from_root(bs, &state_root).unwrap(),
+            &loaded_snapshot.state_tree,
+        );
+
+        let state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height)).unwrap();
```

Comment:
I think this should live inside `read_car` so the caller doesn't need to understand that you were using this tuple. Oh you aren't even deserializing here, so even more so.